lib/ecto/repo/schema.ex

```diff
@@ -116,6 +116,19 @@ defmodule Ecto.Repo.Schema do
     end
 
     header = case query.select do
+      %Ecto.Query.SelectExpr{expr: {:%{}, [], [{:|, _, [{:&, _, [ix]}, args]}]}, fields: fields} ->
```

Comment:
@josevalim There is an issue with taking `select.fields` because the fields have already been dumped. So the second dump will fail with the field has a `:source` option. And removing the second dump means we don't catch `read_only` fields.

The simplest way I can see to address this is to insert something into the field metadata when it is dumped into the planner [here](https://github.com/elixir-ecto/ecto/blob/master/lib/ecto/query/planner.ex#L2242).

Something like:

`{{:., [read_only: true/false], [{:&, [], [ix]}, field]}, [], []}` 

AFAIK this will break some tests but besides that it should be fair game to use the meta like this? We do it for something else [here](https://github.com/elixir-ecto/ecto/blob/master/lib/ecto/query/planner.ex#L1629). 

But please let me know what you think

Reply:
Yeah, I think it's fine to use meta for that!